lib/dataset/sort.cpp

```diff
@@ -21,28 +21,33 @@ template <class T> struct IndicesForSorting {
   static Variable apply(const Variable &key, const SortOrder order) {
     const auto size = key.dims()[key.dim()];
     const auto values = key.values<T>();
-    std::vector<std::pair<T, scipp::index>> pairs;
-    pairs.reserve(size);
-    scipp::index i = 0;
-    for (const auto &value : key.values<T>())
-      pairs.emplace_back(value, i++);
+    std::vector<std::pair<T, scipp::index>> key_index;
+    key_index.reserve(size);
+
+    {
+      scipp::index i = 0;
+      for (const auto &value : key.values<T>())
+        key_index.emplace_back(value, i++);
+    }
 
     if (order == SortOrder::Ascending)
-      std::sort(pairs.begin(), pairs.end(), [](const auto &a, const auto &b) {
-        return nan_sensitive_less(a.first, b.first);
-      });
+      std::sort(key_index.begin(), key_index.end(),
+                [](const auto &a, const auto &b) {
+                  return nan_sensitive_less(a.first, b.first);
+                });
     else
-      std::sort(pairs.begin(), pairs.end(), [](const auto &a, const auto &b) {
-        return nan_sensitive_less(b.first, a.first);
-      });
+      std::sort(key_index.begin(), key_index.end(),
+                [](const auto &a, const auto &b) {
+                  return nan_sensitive_less(b.first, a.first);
+                });
 
     auto indices =
         makeVariable<scipp::index_pair>(Dims{key.dim()}, Shape{size});
-    scipp::index j = 0;
-    for (auto &index : indices.values<scipp::index_pair>().as_span()) {
-      const auto start = pairs[j++].second;
-      index = {start, start + 1};
-    }
+    std::transform(key_index.begin(), key_index.end(),
+                   indices.values<scipp::index_pair>().as_span().begin(),
```

Comment:
```suggestion
                   indices.values<scipp::index_pair>().begin(),
```
?

Reply:
The `as_span` can avoid a potentially more expensive iteration (using `MultiIndex`).

Reply:
ok